tests/test_model.py

```diff
@@ -107,3 +107,73 @@ def test_bfloat16_llama_init(lit_llama, orig_llama) -> None:
 
     out = llama_model2(token_sample.cuda()).float().cpu()
     torch.testing.assert_close(out, expected, atol=5e-3, rtol=1e-3)
+
+
+def copy_adapter_weights(llama_model, orig_llama_model) -> None:
+    # copy the gating parameter
+    for llama_block, orig_llama_block in zip(llama_model.transformer.h, orig_llama_model.layers):
+        if hasattr(llama_block.attn, "gating_factor"):
+            llama_block.attn.gating_factor.copy_(orig_llama_block.attention.gate)
+
+    # In the original model, there is one embedding layer for all blocks combined
+    orig_adapter_wte = orig_llama_model.adapter_query.weight.reshape(
+        orig_llama_model.params.adapter_layer, orig_llama_model.params.adapter_len, orig_llama_model.params.dim
+    )
+
+    # In ours, the embedding layer is split across the individual attention layers
+    index = 0
+    for llama_block in llama_model.transformer.h:
+        if hasattr(llama_block.attn, "adapter_wte"):
+            llama_block.attn.adapter_wte.weight.copy_(orig_adapter_wte[index])
+            index += 1
+
+
+def enable_gate(model):
+    for name, param in model.named_parameters():
+        if "gating_factor" in name or "gate" in name:
+            param.fill_(1)
+
+
+@torch.no_grad()
+def test_adapter_parity(orig_llama_adapter):
+    """Test parity between our implementation of LLaMA-Adapter and the reference code."""
+    import lit_llama.adapter as lit_llama
+    orig_llama = orig_llama_adapter
+    
+    block_size = 32
+    vocab_size = 100
+    n_layer = 2
+    n_head = 4
+    n_embd = 16
+    adapter_prompt_length: int = 10
+    adapter_start_layer: int = 0
+
+    llama_config = lit_llama.LLaMAConfig(
+        block_size=block_size, vocab_size=vocab_size, n_layer=n_layer, n_head=n_head, n_embd=n_embd,
+        adapter_prompt_length=adapter_prompt_length, adapter_start_layer=adapter_start_layer,
+    )
+    orig_llama_config = orig_llama.ModelArgs(
+        dim=n_embd, n_layers=n_layer, n_heads=n_head, vocab_size=vocab_size, norm_eps=1e-5, max_seq_len=block_size,
+        adapter_len=adapter_prompt_length, adapter_layer=(n_layer - adapter_start_layer),
+    )
+
+    batch_size = 3
+    token_sample = torch.randint(
+        0, orig_llama_config.vocab_size, size=(batch_size, orig_llama_config.max_seq_len), dtype=torch.int64
+    )
+
+    llama_model = lit_llama.LLaMA(llama_config)
+    llama_model.apply(llama_model._init_weights)
+    orig_llama_model = orig_llama.Transformer(orig_llama_config)
+
+    copy_weights(llama_model, orig_llama_model)
+    copy_adapter_weights(llama_model, orig_llama_model)
+
+    # make the gate non-zero, otherwise the adapter is disabled and the model
+    # identical to regular LLaMA
+    enable_gate(llama_model)
+    enable_gate(orig_llama_model)
```

Comment:
It took my slow brain a while to realize this fact

Reply:
lol